File: packages/ts/src/helpers/crossvm.ts
/**
 * High-level Cross-VM client — unified calls across QoreChain's three VMs.
 *
 * QoreChain runs EVM, SVM, and CosmWasm side by side and lets a single native
 * account invoke a contract on any of them through the `x/crossvm` module's
 * {@link MsgCrossVMCall}. This helper wraps that message — and the cross-VM query
 * client — so an app developer never hand-builds a `{ typeUrl, value }`, encodes a
 * payload by hand, or remembers a service/method name.
 *
 * The headline capability is {@link CrossVMClient.callAtomic}: pack several
 * `MsgCrossVMCall` messages into ONE transaction body so they execute atomically
 * under a single signature — e.g. an EVM call, an SVM call, and a CosmWasm call
 * that all land together or not at all.
 *
 * A call executes inside the transaction and returns the callee's answer: the
 * result carries `executed`, `data` (the return value) and `gasUsed`, decoded
 * from `MsgCrossVMCallResponse`. Pass `async: true` to queue the call for a
 * later `ProcessQueue` dispatch instead, in which case nothing has run yet and
 * only the message id is meaningful.
 *
 * Per-VM payload encoding (pick exactly one shape per call):
 *  - `{ payload }` — raw bytes / hex, passed through unchanged.
 *  - `{ evm: { abi, functionName, args } }` — ABI-encoded with viem's
 *    `encodeFunctionData` (selector + args). Requires the optional `viem` peer.
 *  - `{ cosmwasm: object }` — `JSON.stringify` then UTF-8 bytes (the CosmWasm
 *    execute-msg convention).
 *  - `{ svm: { data } }` — raw bytes / hex (already an SVM instruction blob).
 *
 * Construct one with {@link createCrossVMClient}, passing a connected
 * {@link TxClient} (for writes) and, optionally, a {@link CrossVmQueryClient}
 * and/or a {@link QorClient} for the reads.
 */

import type { EncodeObject } from "@cosmjs/proto-signing";

import { crossvm as crossvmMsg } from "../messages/qorechain";
import * as crossvmTx from "../codegen/qorechain/crossvm/v1/tx";
import type { TxClient, FeeInput, AutoFeeOptions } from "../tx/builder";
import type { BroadcastResult } from "../tx/broadcast";
import type { CrossVmQueryClient } from "../query/grpc";
import type { QorClient } from "../query/qor";
import type { Coin } from "../query/rest";
import type { QueryMessageResponse } from "../codegen/qorechain/crossvm/v1/query";

/** The three execution environments a cross-VM call can target. */
export type VMType = "evm" | "cosmwasm" | "svm";

/** The set of supported VM type strings. */
export const VM_TYPES = ["evm", "cosmwasm", "svm"] as const satisfies readonly VMType[];

/** A hex string (`0x`-prefixed) as accepted for raw payloads. */
export type Hex = `0x${string}`;

/** Raw, pre-encoded payload — passed through to the chain unchanged. */
export interface RawPayload {
  /** The opaque payload bytes (raw `Uint8Array` or `0x`-hex). */
  payload: Uint8Array | Hex;
}

/** EVM payload built by ABI-encoding a function call with viem. */
export interface EvmPayload {
  evm: {
    /** The contract ABI (viem `Abi`-compatible array). */
    abi: readonly unknown[];
    /** The function to call. */
    functionName: string;
    /** The call arguments. */
    args?: readonly unknown[];
  };
}

/** CosmWasm payload: a JSON execute-message object (stringified to UTF-8). */
export interface CosmWasmPayload {
  /** The CosmWasm execute message (e.g. `{ transfer: { ... } }`). */
  cosmwasm: object;
}

/** SVM payload: a pre-built instruction blob (raw bytes / hex). */
export interface SvmPayload {
  svm: {
    /** The SVM instruction data (raw `Uint8Array` or `0x`-hex). */
    data: Uint8Array | Hex;
  };
}

/** Exactly one of the supported payload shapes. */
export type PayloadInput =
  | RawPayload
  | EvmPayload
  | CosmWasmPayload
  | SvmPayload;

/** Shared write-path options forwarded to {@link TxClient.signAndBroadcast}. */
export interface CrossVMWriteOptions {
  /** Fee: an explicit `StdFee` or `"auto"` (simulate + price). Default `"auto"`. */
  fee?: FeeInput;
  /** Optional memo string. */
  memo?: string;
  /** Auto-fee tuning (gas multiplier / gas price) when `fee` is `"auto"`. */
  autoFee?: AutoFeeOptions;
}

/** Common cross-VM call fields (without the payload or write options). */
export interface CrossVMCallBase {
  /**
   * The VM the call claims to originate from.
   *
   * IGNORED BY THE CHAIN (v3.1.97). The chain derives the origin lane from the
   * execution context rather than trusting the caller's self-description, so
   * setting this changes nothing about how the call is handled. The field is
   * retained — and still sent, defaulting to `"evm"` — purely for wire
   * compatibility, so older nodes and clients keep accepting the message.
   */
  sourceVm?: VMType;
  /** The VM the call targets. */
  targetVm: VMType;
  /** The target contract address/identifier on `targetVm`. */
  targetContract: string;
  /** Optional funds (coins) to forward with the call. */
  funds?: Coin[];
  /**
   * Queue the call for a later `ProcessQueue` dispatch instead of executing it
   * inside this transaction.
   *
   * Defaults to `false`, matching the chain's default: execute now and return
   * the callee's answer (see {@link CrossVMCallResult.data}). With `async: true`
   * the call is only enqueued, so the response carries `executed: false` and no
   * `data` — use it when you do not need the result in this transaction.
   */
  async?: boolean;
}

/** Options for a single cross-VM call (base + payload). */
export type CrossVMCallOptions = CrossVMCallBase & PayloadInput;

/** Options for {@link CrossVMClient.call} (adds write-path options). */
export type CallOptions = CrossVMCallOptions & CrossVMWriteOptions;

/**
 * The decoded `MsgCrossVMCallResponse` fields for one cross-VM call.
 *
 * Populated from the tx's per-message responses when the node returns them
 * (`commit` broadcasts). When they are absent — a `sync`/`async` broadcast, or
 * an older node — `executed` is `false`, `data` is empty and `gasUsed` is `0n`;
 * read the result with {@link CrossVMClient.getMessage} instead.
 */
export interface CrossVMCallOutcome {
  /**
   * Whether the callee actually ran inside this transaction. `false` for a
   * queued (`async: true`) call, whose result is not known yet.
   */
  executed: boolean;
  /** The callee's return value. Empty when the call was queued. */
  data: Uint8Array;
  /** Gas consumed by the callee. `0n` when the call was queued. */
  gasUsed: bigint;
}

/** Result of a single {@link CrossVMClient.call}. */
export interface CrossVMCallResult extends CrossVMCallOutcome {
  /** The cross-VM message id assigned by the chain (parsed from tx events). */
  messageId: string;
  /** The raw broadcast result. */
  result: BroadcastResult;
}

/** Result of an atomic {@link CrossVMClient.callAtomic} batch. */
export interface CrossVMAtomicResult {
  /** The cross-VM message ids assigned by the chain (best-effort, from events). */
  messageIds: string[];
  /**
   * The decoded per-call outcomes, in the order the calls were passed.
   *
   * Empty when the node returned no per-message responses (see
   * {@link CrossVMCallOutcome}).
   */
  outcomes: CrossVMCallOutcome[];
  /** The raw broadcast result for the single packing transaction. */
  result: BroadcastResult;
}

/**
 * Ergonomic client for the `x/crossvm` module.
 *
 * Writes build + sign + broadcast a {@link MsgCrossVMCall}; {@link callAtomic}
 * packs several into one body. {@link buildCall} is the offline build-only path.
 * Reads return the typed query response (or the `qor_` JSON-RPC fallback).
 */
export interface CrossVMClient {
  /** Build, sign, and broadcast a single cross-VM call. */
  call(opts: CallOptions): Promise<CrossVMCallResult>;
  /** Build a single `MsgCrossVMCall` without broadcasting. */
  buildCall(opts: CrossVMCallOptions): EncodeObject;
  /**
   * Pack multiple cross-VM calls into ONE transaction body so they execute
   * atomically under a single signature (the triple-VM headline).
   */
  callAtomic(
    calls: CrossVMCallOptions[],
    opts?: CrossVMWriteOptions,
  ): Promise<CrossVMAtomicResult>;
  /**
   * Read a cross-VM message by id. Uses the typed query client when provided,
   * otherwise falls back to the `qor_getCrossVMMessage` JSON-RPC method.
   */
  getMessage(id: string): Promise<QueryMessageResponse | Record<string, unknown>>;
}

/** Options for {@link createCrossVMClient}. */
export interface CreateCrossVMClientOptions {
  /**
   * Typed query client for {@link CrossVMClient.getMessage} (preferred). Obtain
   * it via {@link connectQueryClients} / {@link createQueryClients}.
   */
  query?: CrossVmQueryClient;
  /**
   * `qor_` JSON-RPC client used as the {@link CrossVMClient.getMessage} fallback
   * (via `qor_getCrossVMMessage`) when no typed query client is supplied.
   */
  qor?: QorClient;
}

const HEX_RE = /^0x[0-9a-fA-F]*$/;

/** Coerce a raw `Uint8Array | Hex` payload to bytes. */
function rawToBytes(data: Uint8Array | Hex): Uint8Array {
  if (typeof data !== "string") return data;
  if (!HEX_RE.test(data)) {
    throw new Error(
      `crossvm: invalid hex payload (expected 0x-prefixed hex, got "${data.slice(0, 12)}...")`,
    );
  }
  const hex = data.slice(2);
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  }
  return bytes;
}

/** UTF-8 encode a JSON object (the CosmWasm execute-msg convention). */
function cosmwasmToBytes(msg: object): Uint8Array {
  return new TextEncoder().encode(JSON.stringify(msg));
}

/**
 * Resolve a {@link PayloadInput} to the raw `Uint8Array` the chain expects.
 *
 * The EVM path lazily imports viem's `encodeFunctionData` so the optional `viem`
 * peer is only required when an `{ evm: ... }` payload is actually used.
 */
async function encodePayload(input: PayloadInput): Promise<Uint8Array> {
  if ("payload" in input) {
    return rawToBytes(input.payload);
  }
  if ("cosmwasm" in input) {
    return cosmwasmToBytes(input.cosmwasm);
  }
  if ("svm" in input) {
    return rawToBytes(input.svm.data);
  }
  // EVM: ABI-encode the call (selector + args) via viem.
  const { encodeFunctionData } = await import("viem");
  const data = encodeFunctionData({
    // viem's Abi typing is structurally compatible; cast at the boundary.
    abi: input.evm.abi as never,
    functionName: input.evm.functionName as never,
    args: (input.evm.args ?? []) as never,
  } as never) as Hex;
  return rawToBytes(data);
}

function requireGetMessageSource(
  query: CrossVmQueryClient | undefined,
  qor: QorClient | undefined,
): void {
  if (!query && !qor) {
    throw new Error(
      "crossvm getMessage requires a query client or a qor client — pass { query } or { qor } to createCrossVMClient",
    );
  }
}

/**
 * Best-effort extraction of cross-VM message ids from a broadcast result.
 *
 * The chain returns the id in `MsgCrossVMCallResponse.messageId`; it is also
 * emitted as a typed event attribute. We scan the broadcast events for any
 * `message_id` / `messageId` attribute so the helper returns ids without an
 * extra round-trip. When none are present, an empty string is returned.
 */
function extractMessageIds(result: BroadcastResult): string[] {
  const events = (result as { events?: unknown }).events;
  const ids: string[] = [];
  if (Array.isArray(events)) {
    for (const ev of events as Array<{ attributes?: unknown }>) {
      const attrs = ev.attributes;
      if (!Array.isArray(attrs)) continue;
      for (const a of attrs as Array<{ key?: unknown; value?: unknown }>) {
        const key = String(a.key ?? "");
        if (key === "message_id" || key === "messageId") {
          ids.push(String(a.value ?? ""));
        }
      }
    }
  }
  return ids;
}

/** The response type URL emitted for a `MsgCrossVMCall`. */
const CROSSVM_CALL_RESPONSE_TYPE_URL =
  "/qorechain.crossvm.v1.MsgCrossVMCallResponse";

/** The outcome reported when the node returned no decodable response. */
function unknownOutcome(): CrossVMCallOutcome {
  return { executed: false, data: new Uint8Array(0), gasUsed: 0n };
}

/**
 * Decode the `MsgCrossVMCallResponse` entries from a broadcast result.
 *
 * Only `commit` broadcasts carry per-message responses; anything else (or an
 * older node) yields an empty list, and callers fall back to
 * {@link CrossVMClient.getMessage}. Entries whose type URL is not a cross-VM
 * call response are skipped, so a mixed transaction body still lines up.
 */
function decodeCallOutcomes(result: BroadcastResult): CrossVMCallOutcome[] {
  const responses = result.msgResponses;
  if (!Array.isArray(responses)) return [];
  const out: CrossVMCallOutcome[] = [];
  for (const r of responses) {
    if (r?.typeUrl !== CROSSVM_CALL_RESPONSE_TYPE_URL) continue;
    try {
      const decoded = crossvmTx.MsgCrossVMCallResponse.decode(r.value);
      out.push({
        executed: decoded.executed,
        data: decoded.data,
        // uint64 is generated as a decimal string; surface it as a bigint.
        gasUsed: BigInt(decoded.gasUsed ?? 0),
      });
    } catch {
      // A response we cannot decode is reported as "unknown" rather than
      // dropped, so positional alignment with the calls is preserved.
      out.push(unknownOutcome());
    }
  }
  return out;
}

/**
 * Create a {@link CrossVMClient} bound to a connected {@link TxClient}.
 *
 * The `TxClient`'s sender address is used as the message `sender`, so the caller
 * never repeats their address. `sourceVm` still defaults to `"evm"` on the wire
 * but is ignored by the chain (see {@link CrossVMCallBase.sourceVm}); `async`
 * defaults to `false`, i.e. execute now and return the callee's answer.
 *
 * @param tx - A connected signing client (from `client.connectTx(signer)`).
 * @param opts - Optional typed query client and/or `qor_` client for reads.
 */
export function createCrossVMClient(
  tx: TxClient,
  opts: CreateCrossVMClientOptions = {},
): CrossVMClient {
  const sender = tx.senderAddress;
  const query = opts.query;
  const qor = opts.qor;

  const buildFrom = (o: CrossVMCallOptions, payload: Uint8Array): EncodeObject =>
    crossvmMsg.crossVmCall({
      sender,
      sourceVm: o.sourceVm ?? "evm",
      targetVm: o.targetVm,
      targetContract: o.targetContract,
      payload,
      funds: o.funds ?? [],
      async: o.async ?? false,
    });

  // Synchronous build-only path. Raw / svm / cosmwasm payloads need no async; the
  // EVM ABI path needs viem, so buildCall pre-encodes EVM payloads inline here.
  const buildCallSync = (o: CrossVMCallOptions, payload: Uint8Array): EncodeObject =>
    buildFrom(o, payload);

  const buildCall = (o: CrossVMCallOptions): EncodeObject => {
    if ("evm" in o) {
      throw new Error(
        "crossvm buildCall: EVM payloads are ABI-encoded asynchronously (viem). " +
          "Use `call`/`callAtomic`, or pre-encode and pass `{ payload }`.",
      );
    }
    // Synchronous encode for raw / cosmwasm / svm.
    let payload: Uint8Array;
    if ("payload" in o) payload = rawToBytes(o.payload);
    else if ("cosmwasm" in o) payload = cosmwasmToBytes(o.cosmwasm);
    else payload = rawToBytes(o.svm.data);
    return buildCallSync(o, payload);
  };

  const call = async (o: CallOptions): Promise<CrossVMCallResult> => {
    const payload = await encodePayload(o);
    const message = buildFrom(o, payload);
    const result = await tx.signAndBroadcast(
      [message],
      o.fee ?? "auto",
      o.memo ?? "",
      { autoFee: o.autoFee },
    );
    const [messageId = ""] = extractMessageIds(result);
    const [outcome = unknownOutcome()] = decodeCallOutcomes(result);
    return { messageId, ...outcome, result };
  };

  const callAtomic = async (
    calls: CrossVMCallOptions[],
    w: CrossVMWriteOptions = {},
  ): Promise<CrossVMAtomicResult> => {
    if (calls.length === 0) {
      throw new Error("crossvm callAtomic: provide at least one call");
    }
    const messages = await Promise.all(
      calls.map(async (o) => buildFrom(o, await encodePayload(o))),
    );
    const result = await tx.signAndBroadcast(
      messages,
      w.fee ?? "auto",
      w.memo ?? "",
      { autoFee: w.autoFee },
    );
    return {
      messageIds: extractMessageIds(result),
      outcomes: decodeCallOutcomes(result),
      result,
    };
  };

  const getMessage = (
    id: string,
  ): Promise<QueryMessageResponse | Record<string, unknown>> => {
    requireGetMessageSource(query, qor);
    if (query) return query.message({ id });
    return qor!.getCrossVmMessage(id);
  };

  return { call, buildCall, callAtomic, getMessage };
}

File: packages/py/src/qorsdk/cross_vm.py
"""High-level unified cross-VM call helper for QoreChain's ``x/crossvm`` module.

QoreChain runs several execution VMs side by side — the EVM, a CosmWasm VM, and
an SVM (Solana-style) VM. The ``x/crossvm`` module routes a call from one VM to a
contract on another as a single message, ``MsgCrossVMCall``:

    MsgCrossVMCall { sender, source_vm, target_vm, target_contract, payload,
                     funds, async }

with type URL ``/qorechain.crossvm.v1.MsgCrossVMCall`` and
``source_vm`` / ``target_vm`` drawn from :data:`VM_TYPES` (``"evm"``,
``"cosmwasm"``, ``"svm"``).

**``source_vm`` is IGNORED by the chain (v3.1.97).** The origin lane is derived
from the execution context — a caller describing itself is not evidence of what
it is — and the field is retained only so the field number stays taken and older
clients that still set it are accepted rather than rejected. The SDK still
accepts and sends it (default ``"evm"``), but setting it has no on-chain effect.

``async_`` (proto field ``async``, 7) queues the call for a later
``MsgProcessQueue`` dispatch instead of running it. The default, ``False``,
executes the call inside the transaction and returns the callee's answer — which
is what a caller that needs the result requires. A queued call's
``MsgCrossVMCallResponse`` carries ``executed=False`` and no ``data`` yet.

:func:`decode_cross_vm_responses` decodes the ``MsgCrossVMCallResponse`` entries
out of a committed transaction's result, surfacing ``message_id`` together with
``executed`` / ``data`` / ``gas_used``.

This helper mirrors the ergonomic rollup/multilayer helpers: it wraps the typed
:func:`qorsdk.messages.qorechain.crossvm.cross_vm_call` composer and the SDK's
tx surface (:func:`~qorsdk.tx.send_messages` / :func:`~qorsdk.tx.build_hybrid_tx`
+ :func:`~qorsdk.tx.broadcast`) so an app can build, sign, and broadcast a
cross-VM call — or several atomically in ONE transaction — without hand-rolling
messages.

Payload encoding (one of three sources, checked in this order):

- ``payload=<bytes>`` — used verbatim (raw VM-native calldata; the only EVM form
  supported here, since EVM ABI-encoding of high-level calls is a TS-only DX
  convenience).
- ``cosmwasm=<dict>`` — ``json.dumps`` UTF-8 encoded (CosmWasm execute message).
- ``svm=<bytes>`` — used verbatim (raw SVM instruction data).

Omitting all three sends an empty payload.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, Literal

from cosmpy.protos.cosmos.base.v1beta1.coin_pb2 import Coin

from .accounts import Secp256k1Account
from .messages._composer import Msg
from .messages.qorechain import crossvm as crossvm_msg
from .pqc import PqcKeypair
from .tx import broadcast, build_hybrid_tx, send_messages

#: The valid cross-VM target/source VM identifiers.
VM_TYPES: tuple[str, ...] = ("evm", "cosmwasm", "svm")

#: A cross-VM VM type: ``"evm"`` | ``"cosmwasm"`` | ``"svm"``.
VmType = Literal["evm", "cosmwasm", "svm"]

#: A Native coin as a plain dict, e.g. ``{"denom": "uqor", "amount": "1000"}``.
CoinDict = dict[str, str]
#: A Native ``StdFee``-shaped dict (as produced by :func:`qorsdk.fees.estimate_fee`).
FeeDict = dict[str, Any]

#: The ``MsgCrossVMCall`` field name for the queue flag. ``async`` is a Python
#: keyword, so the generated protobuf class only accepts it through ``**kwargs``.
_ASYNC_FIELD = "async"

#: Type URL of the response the chain returns for each ``MsgCrossVMCall``.
CROSS_VM_CALL_RESPONSE_TYPE_URL = "/qorechain.crossvm.v1.MsgCrossVMCallResponse"


@dataclass(frozen=True)
class CrossVmCallOptions:
    """One cross-VM call's parameters (shared by ``call`` / ``build_call`` / atomic).

    Exactly one of ``payload`` / ``cosmwasm`` / ``svm`` selects the payload
    bytes; omit all three for an empty payload.
    """

    #: The target VM the call routes to (one of :data:`VM_TYPES`).
    target_vm: VmType
    #: The contract/program address on the target VM.
    target_contract: str
    #: The originating VM. Defaults to ``"evm"``. **Ignored by the chain** since
    #: v3.1.97 (the origin lane is derived from the execution context); still
    #: sent so older nodes keep accepting the message.
    source_vm: VmType = "evm"
    #: Funds to attach, as ``{"denom", "amount"}`` dicts.
    funds: list[CoinDict] | None = None
    #: Raw payload bytes (used verbatim).
    payload: bytes | None = None
    #: A CosmWasm execute message (``json.dumps`` UTF-8 encoded).
    cosmwasm: dict[str, Any] | None = None
    #: Raw SVM instruction data (used verbatim).
    svm: bytes | None = None
    #: Queue the call for a later ``MsgProcessQueue`` dispatch instead of
    #: executing it now. Default ``False`` — execute now and return the answer.
    async_: bool = False


@dataclass(frozen=True)
class CrossVmCallResult:
    """One decoded ``MsgCrossVMCallResponse``.

    Produced by :func:`decode_cross_vm_responses` from a committed transaction's
    result. A queued (``async_=True``) call comes back with ``executed=False``,
    empty ``data`` and ``gas_used=0`` — its answer is only known once
    ``MsgProcessQueue`` dispatches it.
    """

    #: The cross-VM message id, usable with :meth:`CrossVmClient.get_message`.
    message_id: str
    #: ``False`` for a queued call, whose result is not known yet.
    executed: bool
    #: The callee's return value (empty until the call executes).
    data: bytes
    #: Gas the cross-VM execution consumed (``0`` for a queued call).
    gas_used: int


def _validate_vm(vm: str, field: str) -> None:
    if vm not in VM_TYPES:
        raise ValueError(f"{field} must be one of {VM_TYPES!r}, got {vm!r}")


def _resolve_payload(
    payload: bytes | None,
    cosmwasm: dict[str, Any] | None,
    svm: bytes | None,
) -> bytes:
    """Resolve the single payload source into raw bytes (empty when all omitted)."""
    provided = [s for s in (payload, cosmwasm, svm) if s is not None]
    if len(provided) > 1:
        raise ValueError(
            "provide at most one of payload / cosmwasm / svm for the cross-VM payload"
        )
    if payload is not None:
        return payload
    if cosmwasm is not None:
        return json.dumps(cosmwasm, separators=(",", ":")).encode("utf-8")
    if svm is not None:
        return svm
    return b""


def _to_coins(funds: list[CoinDict] | None) -> list[Coin]:
    return [Coin(denom=c["denom"], amount=str(c["amount"])) for c in (funds or [])]


def build_cross_vm_call(
    *,
    sender: str,
    target_vm: VmType,
    target_contract: str,
    source_vm: VmType = "evm",
    funds: list[CoinDict] | None = None,
    payload: bytes | None = None,
    cosmwasm: dict[str, Any] | None = None,
    svm: bytes | None = None,
    async_: bool = False,
) -> Msg:
    """Build a ``MsgCrossVMCall`` :class:`~qorsdk.messages.Msg` (no signing).

    See the module docstring for payload-source resolution. ``source_vm`` and
    ``target_vm`` are validated against :data:`VM_TYPES`.

    :param source_vm: Retained for older nodes; the chain **ignores** it and
        derives the origin lane from the execution context.
    :param async_: Queue the call for ``MsgProcessQueue`` instead of executing
        it in this transaction. Default ``False`` (execute now, return the
        answer).
    """
    _validate_vm(source_vm, "source_vm")
    _validate_vm(target_vm, "target_vm")
    resolved = _resolve_payload(payload, cosmwasm, svm)
    message: Msg = crossvm_msg.cross_vm_call(
        sender=sender,
        source_vm=source_vm,
        target_vm=target_vm,
        target_contract=target_contract,
        payload=resolved,
        funds=_to_coins(funds),
        **{_ASYNC_FIELD: bool(async_)},
    )
    return message


def _result_data_hex(response: Any) -> str:
    """Pull the hex-encoded ``TxMsgData`` out of whatever tx result was passed.

    Accepts a REST broadcast body (``{"tx_response": {...}}``), a bare
    ``tx_response`` dict, an :class:`~qorsdk.track.IncludedTx` (via its ``raw``),
    or the hex string itself. Returns ``""`` when the result carries no data —
    a ``sync``-mode broadcast has not executed yet, so there is nothing to
    decode.
    """
    if response is None:
        return ""
    if isinstance(response, str):
        return response
    raw = getattr(response, "raw", None)
    if raw is not None and not isinstance(response, dict):
        return _result_data_hex(raw)
    if isinstance(response, dict):
        if "tx_response" in response:
            return _result_data_hex(response["tx_response"])
        data = response.get("data")
        if isinstance(data, str):
            return data
    return ""


def decode_cross_vm_responses(response: Any) -> list[CrossVmCallResult]:
    """Decode every ``MsgCrossVMCallResponse`` in a committed tx's result.

    The chain returns the responses in the tx result's ``data`` field, a
    hex-encoded ``TxMsgData`` whose ``msg_responses`` hold one ``Any`` per
    message. Non-cross-VM responses (e.g. from other messages in the same tx)
    are skipped, so an atomic call's results come back in message order.

    Pass the REST broadcast body, a bare ``tx_response`` dict, or an
    :class:`~qorsdk.track.IncludedTx` from :func:`~qorsdk.track.wait_for_tx`.
    A ``sync``-mode broadcast response carries no result yet and yields ``[]``;
    wait for inclusion first.

    :returns: A :class:`CrossVmCallResult` per cross-VM call, in message order.
    """
    from cosmpy.protos.cosmos.base.abci.v1beta1.abci_pb2 import TxMsgData

    from .proto.qorechain.crossvm.v1 import tx_pb2 as crossvm_tx

    data_hex = _result_data_hex(response)
    if not data_hex:
        return []
    try:
        raw = bytes.fromhex(data_hex)
    except ValueError as exc:
        raise ValueError("tx result data is not hex-encoded TxMsgData") from exc
    tx_msg_data = TxMsgData()
    tx_msg_data.ParseFromString(raw)

    results: list[CrossVmCallResult] = []
    for any_msg in tx_msg_data.msg_responses:
        if any_msg.type_url != CROSS_VM_CALL_RESPONSE_TYPE_URL:
            continue
        decoded = crossvm_tx.MsgCrossVMCallResponse()
        decoded.ParseFromString(any_msg.value)
        results.append(
            CrossVmCallResult(
                message_id=decoded.message_id,
                executed=decoded.executed,
                data=bytes(decoded.data),
                gas_used=decoded.gas_used,
            )
        )
    return results


def decode_cross_vm_response(response: Any) -> CrossVmCallResult | None:
    """Decode the FIRST ``MsgCrossVMCallResponse`` in a tx result, if any.

    The single-call companion to :func:`decode_cross_vm_responses`; returns
    ``None`` when the result carries no cross-VM response.
    """
    results = decode_cross_vm_responses(response)
    return results[0] if results else None


class CrossVmClient:
    """Ergonomic build/sign/broadcast client for ``x/crossvm`` cross-VM calls.

    Bind it to a signing :class:`~qorsdk.accounts.Secp256k1Account` plus the tx
    context (chain id, account number, REST url, default fee). The account's
    address is used as the message ``sender``, so callers never repeat it.

    Provide a ``pqc_keypair`` to sign quantum-safe hybrid transactions; otherwise
    a classical secp256k1 tx is built. Pass a ``query`` (a
    :class:`~qorsdk.query.grpc.CrossVmQueryClient`) and/or a ``qor``
    (:class:`~qorsdk.qor.QorClient`) to enable :meth:`get_message`.
    """

    def __init__(
        self,
        *,
        account: Secp256k1Account,
        chain_id: str,
        account_number: int,
        rest_url: str,
        fee: FeeDict,
        sequence: int = 0,
        pqc_keypair: PqcKeypair | None = None,
        query: Any | None = None,
        qor: Any | None = None,
    ) -> None:
        self._account = account
        self._chain_id = chain_id
        self._account_number = account_number
        self._rest_url = rest_url
        self._fee = fee
        self._sequence = sequence
        self._pqc_keypair = pqc_keypair
        self._query = query
        self._qor = qor

    @property
    def sender(self) -> str:
        """The bound account address used as every message's ``sender``."""
        return self._account.address

    def build_call(
        self,
        *,
        target_vm: VmType,
        target_contract: str,
        source_vm: VmType = "evm",
        funds: list[CoinDict] | None = None,
        payload: bytes | None = None,
        cosmwasm: dict[str, Any] | None = None,
        svm: bytes | None = None,
        async_: bool = False,
    ) -> Msg:
        """Build a ``MsgCrossVMCall`` for the bound sender, without broadcasting."""
        return build_cross_vm_call(
            sender=self.sender,
            target_vm=target_vm,
            target_contract=target_contract,
            source_vm=source_vm,
            funds=funds,
            payload=payload,
            cosmwasm=cosmwasm,
            svm=svm,
            async_=async_,
        )

    def _build_msg_from_options(self, opts: CrossVmCallOptions) -> Msg:
        return build_cross_vm_call(
            sender=self.sender,
            target_vm=opts.target_vm,
            target_contract=opts.target_contract,
            source_vm=opts.source_vm,
            funds=opts.funds,
            payload=opts.payload,
            cosmwasm=opts.cosmwasm,
            svm=opts.svm,
            async_=opts.async_,
        )

    def _sign_and_broadcast(
        self,
        messages: list[Msg],
        *,
        sequence: int | None = None,
        memo: str = "",
        mode: str = "sync",
    ) -> Any:
        seq = self._sequence if sequence is None else sequence
        if self._pqc_keypair is not None:
            built = build_hybrid_tx(
                account=self._account,
                pqc_keypair=self._pqc_keypair,
                messages=messages,
                fee=self._fee,
                chain_id=self._chain_id,
                account_number=self._account_number,
                sequence=seq,
                memo=memo,
            )
        else:
            built = send_messages(
                account=self._account,
                messages=messages,
                chain_id=self._chain_id,
                account_number=self._account_number,
                sequence=seq,
                fee=self._fee,
                memo=memo,
            )
        return broadcast(self._rest_url, built.tx_raw_bytes, mode=mode)  # type: ignore[arg-type]

    def call(
        self,
        *,
        target_vm: VmType,
        target_contract: str,
        source_vm: VmType = "evm",
        funds: list[CoinDict] | None = None,
        payload: bytes | None = None,
        cosmwasm: dict[str, Any] | None = None,
        svm: bytes | None = None,
        async_: bool = False,
        sequence: int | None = None,
        memo: str = "",
        mode: str = "sync",
    ) -> Any:
        """Build, sign, and broadcast a single ``MsgCrossVMCall``.

        Returns the decoded broadcast response. See the module docstring for how
        ``payload`` / ``cosmwasm`` / ``svm`` select the payload bytes. Set
        ``async_=True`` to queue the call instead of executing it now.

        Pass the result (or the :class:`~qorsdk.track.IncludedTx` from
        :func:`~qorsdk.track.wait_for_tx`) to
        :func:`decode_cross_vm_response` for the ``message_id`` / ``executed`` /
        ``data`` / ``gas_used`` the chain returned.
        """
        message = self.build_call(
            target_vm=target_vm,
            target_contract=target_contract,
            source_vm=source_vm,
            funds=funds,
            payload=payload,
            cosmwasm=cosmwasm,
            svm=svm,
            async_=async_,
        )
        return self._sign_and_broadcast(
            [message], sequence=sequence, memo=memo, mode=mode
        )

    def call_atomic(
        self,
        options: list[CrossVmCallOptions],
        *,
        sequence: int | None = None,
        memo: str = "",
        mode: str = "sync",
    ) -> Any:
        """Broadcast N cross-VM calls atomically in ONE transaction.

        All ``MsgCrossVMCall`` messages share a single tx, so they succeed or
        fail together. Returns the decoded broadcast response;
        :func:`decode_cross_vm_responses` turns the committed result into one
        :class:`CrossVmCallResult` per call, in message order.

        :raises ValueError: If ``options`` is empty.
        """
        if not options:
            raise ValueError("call_atomic requires at least one cross-VM call")
        messages = [self._build_msg_from_options(o) for o in options]
        return self._sign_and_broadcast(
            messages, sequence=sequence, memo=memo, mode=mode
        )

    def get_message(self, message_id: str) -> Any:
        """Read a cross-VM message by id.

        Prefers the typed ``query`` client's ``Message`` route; falls back to the
        ``qor_getCrossVMMessage`` JSON-RPC method on the ``qor`` client.

        :raises ValueError: If neither a ``query`` nor a ``qor`` client was given.
        """
        if self._query is not None:
            return self._query.message(message_id)
        if self._qor is not None:
            return self._qor.get_cross_vm_message(message_id)
        raise ValueError(
            "get_message requires a query client or a qor client — pass query=… "
            "or qor=… to CrossVmClient"
        )


def create_cross_vm_client(
    *,
    account: Secp256k1Account,
    chain_id: str,
    account_number: int,
    rest_url: str,
    fee: FeeDict,
    sequence: int = 0,
    pqc_keypair: PqcKeypair | None = None,
    query: Any | None = None,
    qor: Any | None = None,
) -> CrossVmClient:
    """Create a :class:`CrossVmClient` bound to a signing account and tx context."""
    return CrossVmClient(
        account=account,
        chain_id=chain_id,
        account_number=account_number,
        rest_url=rest_url,
        fee=fee,
        sequence=sequence,
        pqc_keypair=pqc_keypair,
        query=query,
        qor=qor,
    )


__all__ = [
    "VM_TYPES",
    "CROSS_VM_CALL_RESPONSE_TYPE_URL",
    "VmType",
    "CrossVmCallOptions",
    "CrossVmCallResult",
    "CrossVmClient",
    "build_cross_vm_call",
    "create_cross_vm_client",
    "decode_cross_vm_response",
    "decode_cross_vm_responses",
]
